depthai_ros_driver/include/depthai_ros_driver/ros_agnostic.hpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef DEPTHAI_ROS_DRIVER__ROS_AGNOSTIC_HPP
+#define DEPTHAI_ROS_DRIVER__ROS_AGNOSTIC_HPP
+
+#if defined(USE_ROS2)
+  #include <rclcpp/node.hpp>
+  #define ROS_LOGGER(...) RCUTILS_LOG_WARN_NAMED("depthai_ros", __VA_ARGS__)
+using ROSNodeHandle = std::shared_ptr<rclcpp::Node>;
+using RosTime = rclcpp::Time;
+using RosDuration = rclcpp::Duration;
+#else
+  #include <ros/ros.h>
+  #define ROS_LOGGER(...) ROS_WARN_NAMED("depthai_ros", __VA_ARGS__)
+using ROSNodeHandle = std::shared_ptr<ros::NodeHandle>;
+using RosTime = ros::Time;
+using RosDuration = ros::Duration;
+#endif
+
+namespace rr {
+
+/*
+ * Ros agnostic is a group of ROS wrapper classes, which enables agnostic ROS
+ * implementation of both ros1 and ros2. This can be switched during compile
+ * by defining 'USE_ROS2' in the Cmake file
+ */
+namespace ros_agnostic {
+
+//==============================================================================
+class Publisher
+{
+public:
+  /// @brief Create ros publisher
+  template<class Msg>
+  void create_publisher(
+    ROSNodeHandle node_handle,
+    const std::string& topic_name,
+    const uint32_t queue_size)
+  {
+    #if defined(USE_ROS2)
+    // using qos: KeepLast(QueueSize)
+    _ptr = node_handle->create_publisher<Msg>(topic_name, queue_size);
+    #else
+    _pub = node_handle->advertise<Msg>(topic_name, queue_size);
+    #endif
+  }
+
+  /// @brief Publish ros msg
+  template<typename Msg>
+  void publish(Msg& msg)
+  {
+    #if defined(USE_ROS2)
+    // move the ptr and subscriber check here
+    auto pub = std::static_pointer_cast<rclcpp::Publisher<Msg>>(_ptr);
+    if (pub && pub->get_subscription_count() > 0)
+      pub->publish(msg);
+    #else
+    if (_pub.getNumSubscribers() > 0);
+    _pub.publish(msg);
+    #endif
+  }
+
+private:
+  #if defined(USE_ROS2)
+  std::shared_ptr<void> _ptr;
+  #else
+  ros::Publisher _pub;
+  #endif
+};
```

Comment:
Thoughts on 2 different class implementations, instead of guards everywhere?

Reply:
Since we are switching between ros1  and 2, having the same api class is easier for the user to implement their impl? Anyway, feel free to review again after I tiny up the ros_agnostic impl.